developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,118 @@
+.. _uuid:
+
+UUID Usage in SOF
+#################
+
+We use universally unique identifier (UUID) for component type
+definition in SOF, and for the whole SOF stack, the UUID usage is
+following below rules:
+
+UUID allocation
+***************
+The UUID of a specific component should be specified in the firmware
+component dirver, e.g. for src component, in sof/src/audio/src.c:
+
+.. code-block:: none
+
+   /* c1c5326d-8390-46b4-aa47-95c3beca6550 */
+   DECLARE_SOF_UUID("src", src_uuid, 0xc1c5326d, 0x8390, 0x46b4,
+                    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50);
+
+UUID in topology
+****************
+We should provide UUID string in topology .m4 file in the exactly same
+format as commented in the firmware, e.g for host it is
+c1c5326d-8390-46b4-aa47-95c3beca6550, so in the topology .m4 file
+tools/topology/m4/src.m4:
+
+.. code-block:: none
+
+   dnl Defines the macro for SRC widget
+   define(SRC_UUID_ARRAY, UUID_ARRAY("c1c5326d-8390-46b4-aa47-95c3beca6550"))
+
+We have a .m4 macro UUID_ARRAY() will help to transfer format in #2 (36
+digits) to 16 Bytes byte array, after compiled, the corresponding ALSA
+UUID token defined here will be generated:
+
+.. code-block:: none
+
+   /* vendor tuple for uuid */
+   struct snd_soc_tplg_vendor_uuid_elem {
+           __le32 token;
+           char uuid[16];
+   } __attribute__((packed));
+
+Linux topology driver
+*********************
+The topology driver will parse the 16Bytes UUID token and sent it to the
+firmware in component creation stage, **for all components**.
+
+.. code-block:: none
+
+   /* Generic component tokens */
+   static const struct sof_topology_token comp_tokens[] = {
+           {SOF_TKN_COMP_CORE_ID,
+               SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+               offsetof(struct sof_ipc_comp, core), 0},
+           {SOF_TKN_COMP_UUID,
+               SND_SOC_TPLG_TUPLE_TYPE_UUID, get_token_uuid,
+               offsetof(struct sof_ipc_comp, uuid), 0},
+   };
+
+.. code-block:: none
+
+   static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+                               struct snd_soc_dapm_widget *w,
+                               struct snd_soc_tplg_dapm_widget *tw)
+   {
+   ...
+           ret = sof_parse_tokens(scomp, &template, comp_tokens,
+                                  ARRAY_SIZE(comp_tokens), tw->priv.array,
+                                  le32_to_cpu(tw->priv.size));
+   ...
+   }
+
+UUID arrays stored section
+**************************
+The UUID arrays are stored to SRAM in firmware, though it will lead to a
+1~4KB (depends on how many modules are compiled) memory consumption
+increase. e.g. in src/platform/cannonlake/cannonlake.x.in:
```

Comment:
do we actually need to store the UUID in the firmware?
Wondering if we could optimize this just like we do for strings with traces, the firmware only contains an index and the strings are extracted by indirection. We could do something similar on firmware generation where a specific UUID is replaced by a 'dynamic index', with e.g. the mapping between UUID and index in the firmware manfest. It's kind of clunky if the fimware needs to do tests with 10 bytes to figure out what to do with the IPC.

Reply:
This was actually my question also, That will be the 2nd solution in my previous comment here: https://github.com/thesofproject/sof/pull/2956#issuecomment-631833396

I think I was enlightened during the discussion that this solution might hit issues during dynamic module/component load/unload, the pre-built FW extended manifest might be inconsistent or unaligned between the UUID <-->index pairs. 

@lgirdwood @kv2019i correct me if I am wrong.

Reply:
We're going to hit issues with module load/unload as well for string handling for traces...

Reply:
@plbossart The specific idea you raise matches with the initial design of a UUID dictionary, and it was also implemented, but rejected for SOF, see @lgirdwood 's guidance in https://github.com/thesofproject/sof/pull/2956#issuecomment-631978166 
We have to pick a path one way or the other...

Reply:
@keyonjie where does 1 - 4KB come from ? UUID means we are adding 16 bytes per component.
```grep -rn DECLARE_SOF_UUID --exclude *.lst --exclude *.h | wc -l
35
```
This also includes some duplicates from different platforms so this number is nearer 30 and about 500 bytes. If we are near 1kB or more then it means we rea duplicating UUIDs somewhere.

Reply:
I've found the reason why UUIDs are growing - 69 on master now. They should only be used for components that are loaded via topology, but they seem to being used for kernel subsystems now.  @mmaka1 why are we using UUIDs on kernel subsystems now ? I would expect kernel to
```c
/* system UUIDs are stored in LDC */
DECLARE_SYSTEM_UUID("sys.mem.alloc", "0x1234...........");
```
If userspace needs UUIDs for all kernel objects then it can read them from LDC.

Reply:
Yes, and remember we have name(max 32 Bytes more) stored for each UUID also, we actually store 48 Bytes for each entry, so that is 48 * 69 = 3312B > 3KB.
```
struct sof_uuid_entry {
	struct sof_uuid id;
	const char name[UUID_NAME_MAX_LEN];
};
```

Reply:
> We're going to hit issues with module load/unload as well for string handling for traces...

That's true, I think @mmaka1 is considering changing that for trace also.

Reply:
> Yes, and remember we have name(max 32 Bytes more) stored for each UUID also, we actually store 48 Bytes for each entry, so that is 48 * 69 = 3312B > 3KB.
> 
> ```
> struct sof_uuid_entry {
> 	struct sof_uuid id;
> 	const char name[UUID_NAME_MAX_LEN];
> };
> ```

why would you need to pass the name to the firmware? is it necessary?

Reply:
> > Yes, and remember we have name(max 32 Bytes more) stored for each UUID also, we actually store 48 Bytes for each entry, so that is 48 * 69 = 3312B > 3KB.
> > ```
> > struct sof_uuid_entry {
> > 	struct sof_uuid id;
> > 	const char name[UUID_NAME_MAX_LEN];
> > };
> > ```
> 
> why would you need to pass the name to the firmware? is it necessary?

No, I am here only stating the fact that the whole 48 Bytes are stored in and occupying the SRAM space, see it here:
```
struct sof_uuid_entry {
	struct sof_uuid id;
	const char name[UUID_NAME_MAX_LEN];
};
...
#define DECLARE_SOF_UUID(entity_name, uuid_name,			\
			 va, vb, vc,					\
			 vd0, vd1, vd2, vd3, vd4, vd5, vd6, vd7)	\
	__section(".static_uuids")					\
	static const struct sof_uuid_entry uuid_name = {		\
		{.a = va, .b = vb, .c = vc,				\
		 .d = {vd0, vd1, vd2, vd3, vd4, vd5, vd6, vd7}},	\
		entity_name "\0"					\
	}
```

We are only passing the 16Bytes (corresponding to the 'id' in struct sof_uuid_entry) from driver to FW.

Reply:
Updating offline discussions that:
 - name[UUID_NAME_LAX_LEN] will not be stored
 - only audio components visible in topology will have their uuids stored in FW memory
These both will save the memory footprint.

Reply:
> Updating offline discussions that:
> 
> * name[UUID_NAME_LAX_LEN] will not be stored
> * only audio components visible in topology will have their uuids stored in FW memory
>   These both will save the memory footprint.

Thanks @kv2019i let me update the doc accordingly.

Reply:
This is updated, mark it as resolved to make review easier.